Add mapToList & indexedMapToList Functions

prikhi opened the pull request:
These functions traverse the Array, transform each element, & collect
the results into a List.

This is useful when building HTML from an Array:
```elm
type Exercise
renderExercise : Exercise -> Html msg

renderExercises : Array Exercise -> List (Html msg)
renderExercises =
    mapToList renderExercise
```

prikhi commented:
@jonboiser Mind if I merge this & cut a new release?

jonboiser commented:
@prikhi That's fine with me. Was this function requested in an issue? How will it be useful?

prikhi commented:
> Was this function requested in an issue?

Nah, just going through my projects and pulling out functions I find myself writing all the time.

> How will it be useful?

I use it when rendering an array of items into HTML:
```elm
type SomeItem = ...

renderItem : SomeItem -> Html msg

itemList : Array SomeItem -> Html msg
itemList items =
    Html.ul [] <| Array.mapToList renderItem items
```

That way I don't do a double pass over all the items like with `Array.toList >> List.map renderingFunction` or `Array.map renderingFunction >> Array.toList`.

jonboiser commented on src/Array/Extra.elm:
```diff
@@ -137,6 +137,27 @@ apply fs xs =
+
+
+{-| Apply a function to the array with the index as the first argument,
+collecting the results in a List. 
```
Could you add an example for this function too?

jonboiser commented:
Sounds good.

I'm having a little trouble with the names, though. I had to read the code to understand that `mapToList` meant something like 'map (over an array) then convert to a list'. But initially I read it as something like 'convert a map to a list'.

Do you have any other ideas for better names.

prikhi commented:
> initially I read it as something like 'convert a map to a list'

Ah, I assumed it'd be evident since I usually do something like this:

```elm
import Array exposing (Array)
import Array.Extra as Array

renderThem : Array Something -> Html msg
renderThem = div [] <| Array.mapToList (\s -> text s.thing)
```

> Do you have any other ideas for better names

`buildList`, `foldIntoList`, `foldToList`, `toListWith`?

jonboiser commented:
I don't have any suggestions.

I think this PR is good to merge as soon as you add an example in the docstring for indexedMapToList

prikhi commented:
added the example, fixed the generated list being in the opposite order of the array, & rebased/squashed the commits.

might make an additional PR to add some tests.

jonboiser commented:
@prikhi Sounds good! Could you publish after this is merged.
